pkg/mimir/modules.go

```diff
@@ -220,6 +222,15 @@ func (t *Mimir) initSanityCheck() (services.Service, error) {
 }
 
 func (t *Mimir) initServer() (services.Service, error) {
+	// This allows us to reject gRPC requests early -- before they are fully read into memory, only based on method name.
+	t.Cfg.Server.GRPCOptions = append(t.Cfg.Server.GRPCOptions, grpc.InTapHandle(func(ctx context.Context, info *tap.Info) (context.Context, error) {
+		var err error
+		if info.FullMethodName == "/cortex.Ingester/Push" && t.Ingester != nil {
+			err = t.Ingester.AcceptPushRequest()
+		}
+		return ctx, err
```

Comment:
if the error here isn't a gRPC status error the library will automatically make the status code PERMISSION_DENIED. I wonder if we shouldn't wrap this with something to keep the existing behaviour. I'm not 100% sure whether we return UNAVAILABLE or INTERNAL when the instance limit is reached

Reply:
All errors returned by `ingester.AcceptPushRequest()` are already convertible via `status.FromError` (all instance limit errors are, and similarly error returned by `i.checkRunning()`. Tests check for it too. I will add comments explaining this.